java/memory/memory-netty/src/test/java/org/apache/arrow/memory/TestForeignAllocation.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import org.apache.arrow.memory.util.MemoryUtil;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestForeignAllocation {
+  BufferAllocator allocator;
+
+  @Before
+  public void before() {
+    allocator = new RootAllocator();
+  }
+
+  @After
+  public void after() {
+    allocator.close();
+  }
+
+  @Test
+  public void wrapForeignAllocation() {
+    final long bufferSize = 16;
+    UnsafeForeignAllocationManager allocationManager = new UnsafeForeignAllocationManager(allocator, bufferSize);
+    try {
+      assertEquals(0, allocator.getAllocatedMemory());
+      ArrowBuf buf = allocator.wrapForeignAllocation(allocationManager);
```

Comment:
Although the code already works greatly I think we may have chance to optimize such usages by decoupling BufferAllocator from AllocationManager? @lwhite1

Reply:
> It would be nice to avoid the double reference count but I think that will require some more intricate surgery on the allocator internals.

(Sorry for asking more questions on an already approved PR)

Regarding the double reference counting, I am not sure about why it's actually needed? In the previous PR we didn't actually touch anything about ref counting and things went well.

Reply:
It didn't actually implement the API. It just suppressed an exception.

Reply:
`new AllocationManager(allocator)` creates an zero ref count owning ledger. I guess that ledger may lead to unexpected behavior when user call `wrapForeignAllocation(...)` from another allocator. Maybe decoupling allocator from AllocationManager's constructor (not the manager itself) will be a solution.

Reply:
Ah, ok, thanks. I think I see now. I'll give that a test, and I think I might change it so that the ForeignAllocation is _only_ a release callback, pointer, and size. That way the details can all be properly encapsulated.